File: py/aoc2023/day22.py
"""
Day 22: Sand Slabs
"""

from collections import defaultdict

SAMPLE_INPUT = """
1,0,1~1,2,1
0,0,2~2,0,2
0,2,3~2,2,3
0,0,4~0,2,4
2,0,5~2,2,5
0,1,6~2,1,6
1,1,8~1,1,9
"""


# pylint: disable=too-many-locals,redefined-outer-name
def _parse(data):
    bricks = [
        ([int(num) for num in first], [int(num) for num in second])
        for line in data.splitlines()
        if len(parts := line.split("~", maxsplit=1)) == 2
        and len(first := parts[0].split(",", maxsplit=2)) == 3
        and len(second := parts[1].split(",", maxsplit=2)) == 3
        and all(num.isdecimal() for num in first + second)
    ]
    bricks.sort(key=lambda brick: brick[0][2])
    heights = defaultdict(int)
    for first, second in bricks:
        (x0, y0, z0), (x1, y1, z1) = first, second
        dz = (
            max(heights[(x, y)] for x in range(x0, x1 + 1) for y in range(y0, y1 + 1))
            + 1
            - z0
        )
        first[2], second[2] = z0, z1 = z0 + dz, z1 + dz
        for x in range(x0, x1 + 1):
            for y in range(y0, y1 + 1):
                heights[(x, y)] = z1
    bricks.sort(key=lambda brick: brick[0][2])
    rdeps, deps = defaultdict(set), defaultdict(set)
    for i, ((x0, y0, _), (x1, y1, z1)) in enumerate(bricks):
        for j, ((x2, y2, z2), (x3, y3, _)) in enumerate(bricks[i + 1 :]):
            if z2 <= z1:
                continue
            if z2 > z1 + 1:
                break
            if x0 <= x3 and x2 <= x1 and y0 <= y3 and y2 <= y1:
                rdeps[i].add(i + j + 1)
                deps[i + j + 1].add(i)
            j += 1
    return len(bricks), rdeps, deps


def part1(data):
    """
    >>> part1(SAMPLE_INPUT)
    5
    """
    n, _, deps = _parse(data)
    return n - len({next(iter(below)) for below in deps.values() if len(below) == 1})


def part2(data):
    """
    >>> part2(SAMPLE_INPUT)
    7
    """
    _, rdeps, deps = _parse(data)
    acc = 0
    for item in rdeps.items():
        _deps = {above: set(belows) for above, belows in deps.items()}
        stack = [item]
        while stack:
            below, aboves = stack.pop()
            for above in aboves:
                belows = _deps[above]
                if below not in belows:
                    continue
                belows.remove(below)
                if belows:
                    continue
                acc += 1
                stack.append((above, rdeps.get(above, ())))
    return acc


parts = (part1, part2)

File: rs/src/day22.rs
use rayon::iter::{IntoParallelRefIterator, ParallelIterator};
use std::collections::btree_map::Entry;
use std::collections::{BTreeMap, BTreeSet};

type Multimap<T> = BTreeMap<T, BTreeSet<T>>;

fn parse(data: &str) -> Option<(usize, Multimap<usize>, Multimap<usize>)> {
    let mut bricks = data
        .lines()
        .map(|line| {
            let (first, second) = line.split_once('~')?;
            {
                let mut iter = first.splitn(3, ',');
                Some((
                    iter.next()?.parse::<i32>().ok()?,
                    iter.next()?.parse::<i32>().ok()?,
                    iter.next()?.parse::<i32>().ok()?,
                ))
            }
            .zip({
                let mut iter = second.splitn(3, ',');
                Some((
                    iter.next()?.parse::<i32>().ok()?,
                    iter.next()?.parse::<i32>().ok()?,
                    iter.next()?.parse::<i32>().ok()?,
                ))
            })
        })
        .collect::<Option<Vec<_>>>()?;
    bricks.sort_by_key(|((_, _, z), _)| *z);
    let mut heights = BTreeMap::new();
    for ((x0, y0, z0), (x1, y1, z1)) in &mut bricks {
        let z = (*x0..=*x1)
            .flat_map(|x| {
                let heights = &heights;
                (*y0..=*y1).map(move |y| heights.get(&(x, y)).copied().unwrap_or(0))
            })
            .max()?;
        (*z0, *z1) = (z + 1, z + 1 - *z0 + *z1);
        let z = *z1;
        heights.extend((*x0..=*x1).flat_map(|x| (*y0..=*y1).map(move |y| ((x, y), z))));
    }
    bricks.sort_by_key(|((_, _, z), _)| *z);
    let (mut rdeps, mut deps) = (Multimap::new(), Multimap::new());
    for (i, ((x0, y0, _), (x1, y1, z1))) in bricks.iter().enumerate() {
        for (j, ((x2, y2, _), (x3, y3, _))) in bricks
            .iter()
            .enumerate()
            .skip(i + 1)
            .skip_while(|(_, ((_, _, z2), _))| *z2 <= *z1)
            .take_while(|(_, ((_, _, z2), _))| *z2 == *z1 + 1)
        {
            if *x0 <= *x3 && *x2 <= *x1 && *y0 <= *y3 && *y2 <= *y1 {
                rdeps.entry(i).or_insert_with(BTreeSet::new).insert(j);
                deps.entry(j).or_insert_with(BTreeSet::new).insert(i);
            }
        }
    }
    Some((bricks.len(), rdeps, deps))
}

pub fn part1(data: &str) -> Option<usize> {
    let (size, _, deps) = parse(data)?;
    Some(
        size - deps
            .values()
            .filter_map(|values| {
                let mut iter = values.iter();
                iter.next().filter(|_| iter.next().is_none())
            })
            .collect::<BTreeSet<_>>()
            .len(),
    )
}

pub fn part2(data: &str) -> Option<usize> {
    let (_, rdeps, deps) = parse(data)?;
    Some(
        rdeps
            .par_iter()
            .map(|item| {
                let mut sum = 0usize;
                let mut deps = deps.clone();
                let mut stack = vec![item];
                while let Some((below, aboves)) = stack.pop() {
                    for above in aboves {
                        let Entry::Occupied(mut entry) = deps.entry(*above) else {
                            continue;
                        };
                        if entry.get_mut().remove(below) && entry.get().is_empty() {
                            sum += 1;
                            entry.remove();
                            if let Some(above2) = rdeps.get(above) {
                                stack.push((above, above2));
                            }
                        }
                    }
                }
                sum
            })
            .sum(),
    )
}

#[cfg(test)]
mod tests {
    use super::*;
    use indoc::indoc;
    use pretty_assertions::assert_eq;

    static EXAMPLE: &str = indoc! {"
        1,0,1~1,2,1
        0,0,2~2,0,2
        0,2,3~2,2,3
        0,0,4~0,2,4
        2,0,5~2,2,5
        0,1,6~2,1,6
        1,1,8~1,1,9
    "};

    #[test]
    fn part1_examples() {
        assert_eq!(Some(5), part1(EXAMPLE));
    }

    #[test]
    fn part2_examples() {
        assert_eq!(Some(7), part2(EXAMPLE));
    }
}
